Assets/Mirror/Tests/Editor/QuaternionCompressionTest.cs

```diff
@@ -0,0 +1,146 @@
+using System.Collections.Generic;
+using NUnit.Framework;
+using UnityEngine;
+
+namespace Mirror.Tests
+{
+    public class QuaternionCompressionTest
+    {
+        // worse case where xyzw all equal error in largest is ~1.732 times greater than error in smallest 3
+        // High/Low Precision fails when xyzw all equal,
+        // *1.8f is enough to allow extra error in largest,
+        internal const float HighestPrecision = 0.000000169f * 1.8f;
+        internal const float MediumPrecision = 0.00138f;
+        // *1.2f is enough to allow extra error in largest
+        internal const float LowPrecision = 0.0110f * 1.2f;
+
+        [Test]
+        [TestCaseSource(nameof(GetTestCases))]
```

Comment:
is this common practice for unit tests?
it seems very over engineered, compared to the old, boring way of simply having 1-2 test functions per function that you want to test :)

Reply:
This is quite common vis,  this is called test fixtures,  you have one tests,  and you repeat the test with many different values instead of repeating that code over and over.

Much better than having one large test,  or having 20 tests that do the same thing

Reply:
Its best practice to have 1 test case for 1 thing you are testing so that if 1 thing breaks you know exactly what it is.

With NUnit you can use `TestCase` if you have args that can be constant (like int/string/etc) but for other types you need to use `TestCaseSource` and return a collection of cases. For different use case the test function can even return a value and the TestCaseSource can check that value using `TestCaseData(...).Returns`

In the test runner it shows up like this so if 1 case files you can see exactly which one it is
![image](https://user-images.githubusercontent.com/23101891/97332568-7f89ae80-1872-11eb-98d1-9da940c56756.png)

There is docs on TestCaseSource here and different ways to use it https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcasesource.html